src/iceberg/transform.h

```diff
@@ -56,14 +56,45 @@ enum class TransformType {
   kVoid,
 };
 
+/// \brief Get the relative transform name
+constexpr std::string_view ToString(TransformType type) {
+  switch (type) {
+    case TransformType::kUnknown:
+      return "unknown";
+    case TransformType::kIdentity:
+      return "identity";
+    case TransformType::kBucket:
+      return "bucket";
+    case TransformType::kTruncate:
+      return "truncate";
+    case TransformType::kYear:
+      return "year";
+    case TransformType::kMonth:
+      return "month";
+    case TransformType::kDay:
+      return "day";
+    case TransformType::kHour:
+      return "hour";
+    case TransformType::kVoid:
+      return "void";
+    default:
+      return "invalid";
+  }
+}
+
 /// \brief A transform function used for partitioning.
 class ICEBERG_EXPORT TransformFunction : public util::Formattable {
  public:
-  explicit TransformFunction(TransformType type);
+  explicit TransformFunction(TransformType transform_type,
+                             std::shared_ptr<Type> source_type);
   /// \brief Transform an input array to a new array
   virtual expected<ArrowArray, Error> Transform(const ArrowArray& data) = 0;
   /// \brief Get the transform type
-  virtual TransformType transform_type() const;
+  TransformType transform_type() const;
+  /// \brief Get the source type of transform function
+  std::shared_ptr<Type> source_type() const;
+  /// \brief Get the result type of transform function
+  virtual expected<std::shared_ptr<Type>, Error> result_type() const = 0;
```

Comment:
`snake_case` is only for trivial getters. This should either be `shared_ptr<Type> const& result_type() const` or `expected<...> ResultType() const`.